book/src/proptest-derive/errors.md

```diff
@@ -0,0 +1,631 @@
+}
+```
+
+## E0009
+
+This error happens if `#[proptest(weight = <integer>)]` is applied to an item
+where this does not make sense, such as a struct field. For example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Foo {
+    a: u32,
+    #[proptest(weight = 42)]
+    b: u32,
+}
+```
+
+The `weight` attribute only is sensible where proptest has a choice between
+multiple items, such as enum variants. In contrast, with struct fields proptest
+must provide a value for _every_ field so there is no "this-or-that" choice.
+
+## E0010
+
+This error occurs if `#[proptest(params = "type")]` and/or
+`#[proptest(no_params)]` are set on both an item and its parent.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+#[proptest(params = "String")]
+struct Foo {
+    #[proptest(no_params)]
+    bar: String,
+}
+```
+
+If the parent item has any explicit parameter configuration, it totally defines
+the parameters for the whole `Arbitrary` implementation and the child items
+must work with that and cannot specify their own parameters.
+
+## E0011
+
+This error occurs if `#[proptest(params = "type")]` is set on a field but no
+explicit strategy is configured with `#[proptest(strategy = "expr")]`. For
+example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Foo {
+    #[proptest(param = "u8")]
+    some_string: String,
+}
+```
+
+This example illustrates why both must be specified: `String`'s arbitrary
+implementation takes a `proptest::string::StringParam`, but here we try to pass
+it a `u8`.
+
+While the generated code could work if the type given by `param` is the same as
+that for the default strategy, there would be no purpose in specifying the
+parameter type by hand; therefore specifying only `param` is in all cases
+forbidden.
+
+## E0012
+
+This error occurs if `#[proptest(filter = "expr")]` is set on an item, but the
+item containing it specifies a direct way to generate the whole value, which
+would thus occur without consulting the filter.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum Foo {
+    #[proptest(value = "Foo::V1(42)")]
+    V1 {
+        #[proptest(filter = "is_even")]
+        field: u32,
+    },
+    // ...
+}
+```
+
+In this example, the entire `V1` variant specifies how to generate itself
+wholesale. As a result, the `filter` clause on `field` has no opportunity to
+run.
+
+## E0013
+
+This error would occur if an outer attribute of the form `#![proptest(..)]`
+were applied to something underneath a `#[derive(Arbitrary)]`.
+
+As of Rust 1.30.0, there are no known ways to produce this error since the Rust
+compiler will reject the attribute first.
+
+## E0014
+
+This error occurs if a bare `#[proptest]` attribute is applied to anything.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Foo {
+    #[proptest]
+    field: u8,
+}
+```
+
+The only legal use of the attribute is the form `#[proptest(..)]`.
+
+## E0015
+
+This error occurs if an attribute of the form `#[proptest = value]` is
```

Comment:
Aside: I've wondered if we should give this attribute form meaning...
Maybe it should be synonymous with `#[proptest(strategy = expr)]` but that is perhaps not super clear;
On enum variants, I would also expect this to be for weights...

Reply:
That syntax would feel unclear to me. Let's keep it explicit for now, we can always add some meaning to it later.

Reply:
Yeah I agree 👍